generator/validate/gen.go

```diff
@@ -0,0 +1,97 @@
+package interfaces
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/devfile/api/generator/genutils"
+	"github.com/iancoleman/strcase"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/helpgen generate:headerFile=../header.go.txt,year=2020 paths=.
+
+// +controllertools:marker:generateHelp
+
+// Generator validates the consistency of the API GO code.
+//
+// Validaty check are related to unions, patchStrategy, and optional fields.
+type Generator struct{}
+
+// RegisterMarkers registers the markers of the Generator
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	err := genutils.RegisterUnionMarkers(into)
+	if err != nil {
+		return err
+	}
+	return crdmarkers.Register(into)
+}
+
+// Generate generates the artifacts
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	for _, root := range ctx.Roots {
+		ctx.Checker.Check(root, func(node ast.Node) bool {
+			// ignore interfaces
+			_, isIface := node.(*ast.InterfaceType)
+			return !isIface
+		})
+
+		root.NeedTypesInfo()
+
+		packageTypes := map[string]*markers.TypeInfo{}
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			packageTypes[info.Name] = info
+		}); err != nil {
+			root.AddError(err)
+			return nil
```

Comment:
Only adding the error to root allows the other generators to run (when several generators are run sequentially in one command line). The errors are output at the end of the process.
When returning the error in the generator itself, I assume it would stop the command line here.

This doesn't really make a difference in our case since we use a single generator in each command line run.